File: app/src/main/java/com/example/test/mvvmsampleapp/viewmodel/ProjectViewModel.java
package com.example.test.mvvmsampleapp.viewmodel;

import android.app.Application;
import android.arch.lifecycle.AndroidViewModel;
import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.MutableLiveData;
import android.arch.lifecycle.Transformations;
import android.databinding.ObservableField;
import android.support.annotation.NonNull;
import android.util.Log;

import com.example.test.mvvmsampleapp.service.model.Project;
import com.example.test.mvvmsampleapp.service.repository.ProjectRepository;

import javax.inject.Inject;

public class ProjectViewModel extends AndroidViewModel {
    private static final String TAG = ProjectViewModel.class.getName();
    private static final MutableLiveData ABSENT = new MutableLiveData();
    {
        //noinspection unchecked
        ABSENT.setValue(null);
    }

    private final LiveData<Project> projectObservable;
    private final MutableLiveData<String> projectID;

    public ObservableField<Project> project = new ObservableField<>();

    @Inject
    public ProjectViewModel(@NonNull ProjectRepository projectRepository, @NonNull Application application) {
        super(application);

        this.projectID = new MutableLiveData<>();

        projectObservable = Transformations.switchMap(projectID, input -> {
            if (input.isEmpty()) {
                Log.i(TAG, "ProjectViewModel projectID is absent!!!");
                return ABSENT;
            }

            Log.i(TAG,"ProjectViewModel projectID is " + projectID.getValue());

            return projectRepository.getProjectDetails("Google", projectID.getValue());
        });
    }

    public LiveData<Project> getObservableProject() {
        return projectObservable;
    }

    public void setProject(Project project) {
        this.project.set(project);
    }

    public void setProjectID(String projectID) {
        this.projectID.setValue(projectID);
    }
}

File: app/src/main/java/com/example/test/mvvmsampleapp/viewmodel/ProjectViewModel.kt
package com.example.test.mvvmsampleapp.viewmodel

import android.app.Application
import android.arch.lifecycle.AndroidViewModel
import android.arch.lifecycle.LiveData
import android.arch.lifecycle.MutableLiveData
import android.arch.lifecycle.Transformations
import android.databinding.ObservableField
import android.util.Log

import com.example.test.mvvmsampleapp.service.model.Project
import com.example.test.mvvmsampleapp.service.repository.ProjectRepository

import javax.inject.Inject

class ProjectViewModel @Inject
constructor(projectRepository: ProjectRepository,
            application: Application) : AndroidViewModel(application) {

    val observableProject: LiveData<Project>
    private var projectID: MutableLiveData<String> = MutableLiveData()

    var project = ObservableField<Project>()

    companion object {

        private val TAG = ProjectViewModel::class.java.name
        private val ABSENT = MutableLiveData<Project>()

        init {
            ABSENT.value = null
        }
    }

    init {
        observableProject = Transformations.switchMap(projectID) { input ->

            if (input.isEmpty()) {
                Log.i(TAG, "ProjectViewModel projectID is absent!!!")
                return@switchMap ABSENT
            }

            Log.i(TAG, "ProjectViewModel projectID is " + projectID.value)
            projectRepository.getProjectDetails("Google", projectID.value)
        }
    }

    fun setProject(project: Project) {
        this.project.set(project)
    }

    fun setProjectID(projectID: String?) {
        this.projectID.value = projectID
    }
}
